### PR TITLE
Test the exclude option on snapshot create/decode

### LINKED ISSUE #82: Tests for store/clone/restore excludes
Exclude patterns need tests.

### PR DESCRIPTION
**Pull request by penguwin**
Testing the exclude option in `TestSnapshotCreate` should bring cover the following
commands:
- store
- restore
- clone

This pr should resolve #82

### CODE REVIEW

**Comment by muesli** [comment]
This needs to be a bit more thorough:

1. We need the test to exclude certain files and then verify the snapshot really doesn't contain them.
2. We need a snapshot with certain files included, that will then be excluded during a restore. Verify they really haven't been restored.

**Comment by penguwin** [comment]
I've now have worked on the code to match the mentioned points by @muesli , yet I'm a bit unsatisfied with my code as it seems a bit messy to me - so feel free to mention further improvements.

File: snapshot_test.go [hunk 1]
```diff
@@ -116,6 +119,16 @@ func TestSnapshotCreate(t *testing.T) {
 				return
 			}
 
+			// Check if we have excluded all the paths in tt.excludes
+			for _, v := range snapshot.Archives {
+				for _, e := range tt.excludesStore {
+					if e == v.Path {
+						t.Errorf("Failed excluding specified files from snapshot store operation: %s", e)
+						return
```
Comment (muesli) [inline snapshot_test.go:127]: We should probably just log the error and `continue` instead of `return` here.

File: snapshot_test.go [hunk 2]
```diff
@@ -166,6 +179,29 @@ func TestSnapshotCreate(t *testing.T) {
...
+					excludedFile := filepath.Join(targetdir, excl)
+
+					if _, err := os.Stat(excludedFile); os.IsNotExist(err) {
+						if excludedFile == file1 {
+							isExcluded = true
+							break
+						}
+					} else if err != nil { // Some other error occured
+						t.Errorf("Failed to stat file %s: %v", excl, err)
+						return
```
Comment (muesli) [inline snapshot_test.go:195]: `continue` instead of `return`.

File: snapshot_test.go [hunk 3]
```diff
@@ -166,6 +179,29 @@ func TestSnapshotCreate(t *testing.T) {
...
+					} else { // File probably exists
+						t.Errorf("Failed to exclude file %s while decoding snapshot", excl)
+						return
```
Comment (muesli) [inline snapshot_test.go:198]: `continue` instead of `return`.

Comment (muesli) [reply snapshot_test.go:198]: We could drop this else block, as the above conditions either break or continue/return.

File: snapshot_test.go [hunk 4]
```diff
@@ -37,17 +38,19 @@ func TestSnapshotCreate(t *testing.T) {
...
-		{CompressionNone, 1},
-		{CompressionGZip, 1},
+		{CompressionNone, 0, []string{"snapshot.go"}, []string{}},
+		{CompressionFlate, 0, []string{"snapshot_test.go"}, []string{}},
+		{CompressionGZip, 0, []string{"snapshot.go", "snapshot_test.go"}, []string{}},
+		{CompressionLZMA, 0, []string{}, []string{"snapshot.go"}},
+		{CompressionZlib, 0, []string{}, []string{"snapshot_test.go"}},
+		{CompressionZstd, 0, []string{}, []string{"snapshot.go", "snapshot_test.go"}},
+		{CompressionNone, 1, []string{"snapshot.go"}, []string{"snapshot_test.go"}},
+		{CompressionGZip, 1, []string{}, []string{}},
```
Comment (muesli) [inline snapshot_test.go:53]: I'd like to keep the original compression tests without any excludes, and then add the exclude tests separately, once with `CompressionNone` and once with `CompressionGZip`.

**Comment by penguwin** [comment]
Applied the changes as requested.
Also I've now switched over to https://github.com/muesli/combinator for generating all possible combinations which we want to test (as we've discussed)

**Comment by penguwin** [comment]
Bumped as the logs have expired

**Review by muesli — APPROVED** [review]

**Pull request merged**